hail/build.gradle

```diff
@@ -204,6 +204,9 @@ test {
 
     // make poop emoji work in generated bytecode
     systemProperty "file.encoding", "utf-8"
+    
+    // Enable C++ codegen for tests
+    environment HAIL_ENABLE_CPP_CODEGEN: '1'
```

Comment:
I don't think the primary tests should use C++ until most users/deployments are using C++.  Can we run a subset of tests a second time with C++ enabled?  Maybe UnsafeSuite, AnnotationSuite.testReadWrite in Scala, and test.hail.matrixtable.test_file_formats in Python?  Maybe something that uses joins, like JoinSuite in Scala?

Reply:
Alternatively, we could just run the full tests twice while we wait while we wait for the deployment story to get fixed so we can enable this by default.  I'm also working with Dan to get the CI to run different parts of the tests in parallel, although it will be a couple weeks before that happens, at least.

Reply:
It's certainly possible to rewrite it in the every-byte-access-looks-like-a-call style.  It's also possible that would make it significantly slower, because all those calls would perturb register allocation.  And there are decisions about which calls to have as inlined methods, and which to do as real procedure calls.  In the Scala/Java world, the JVM is going to take its own decisions regardless of how you code it, but in  C++ the programmer has to take a stab at it (the compiler may decide to *not* inline a possibliy-inline call, but it will has now way to take a definitely-not-inline call and flip it to inline).

As an aside, compilation with -O2 has got a lot slower with the change from having a 
combination of member variables and local variable, to making all variables local.   Simplification can have unpredictable consequences.

Reply:
Does anyone know a good way to run a subset of tests that isn't just a single suite ?

Reply:
> It's certainly possible to rewrite it in the every-byte-access-looks-like-a-call style. It's also possible that would make it significantly slower, because all those calls would perturb register allocation.

I don't understand.  Everything should be inline up to and including the check that the buffer is not empty, and the call to fill the buffer in any case is going to be a real call.

Reply:
From the tuning done back in August, the fastest version for chr22 would handle the
integer-single-byte-encoding with an inline method (and with a LIKELY/UNLIKELY hint,
and a carefully-chosen sentinel value so the likely path had only 1 test+branch), but would 
make a procedure call for the multi-byte encodings.

It's important for the frequently-executed paths to be inline.  Having infrequently-executed code inline is ok up to a point, but may confuse the g++clang register allocator in ways which
hurt the frequently-executed paths.  And could hurt I-cache locality.
Also a huge amount of expanded inline code may produce a very large
g++/clang IR for the procedure, which is then slow to compile (especially with optimization).

Inline-everything is not necessarily optimal.

Reply:
There are lots of ways to do this.  What you really want - and what AFAIK no current language supports in any nice way - is a special slow-path procedure call, which preserves all the register state for the fast-path by saving/restoring all registers around the slow-path call.

The current code attempts to achieve a similar effect by doing "goto pull;" which then does the procedure call, and returns through the switch-and-goto to a place where the values of local variables are consistent with the scan position (often going through some reinitialization
and rollback).

The apparent complexity of "if (!this->decode_int(addr) goto pull;" should be quite easy
for the compiler to optimize when decode_int is an inline method.  Though with compilers, who knows ?  The anecdotal evidence is that the code written this way holds the record so far.

Anyhow, let's get it committed.  Someone's going to redesign the encodings and file format soon anyway, so that would be the time to refactor and re-tune.